### PR TITLE
[v6r17] Return directly status in se.getStatus()

### PR DESCRIPTION
**Pull request by phicharp**
BEGINRELEASENOTES
* DMS
CHANGE: as se.getStatus() always returns S_OK, return directly the values. Remove the checks of OK everywhere
* TS
FIX TransformationDB.py: set the ExternalID before the ExternalStatus in order to avoid inconsistent tasks if setting the ExternalID fails
* StorageManagementSystem
FIX StorageManagementClient.py: return the full list of onlineSites while it was previously happy with only one

ENDRELEASENOTES

### CODE REVIEW

**Comment by andresailer** [comment]
This is the kind of interface change that should not be done in a patch release.

**Comment by phicharp** [comment]
I have some sympathy for your comment ;-), but since we are back to the "one release a year" pace, this is the change that never gets done... We can move it to Integration of course, but I won't be there anymore until it makes it to production I am afraid...
I can make as was made at several other places which is to return both structures as internally DIRAC is OK and VO extensions would have enough time to adapt... Is this agreeable?

**Comment by phicharp** [comment]
To be clear:
`retDict.update( S_OK(retDict))`

**Comment by phicharp** [comment]
or safer
`retDict.update( S_OK(retDict.copy()))`

**Comment by andresailer** [comment]
Yes, that would work. Thanks for your sympathy :)

Would also be nice to have the return type documented in the docstring

**Comment by phicharp** [comment]
Would you complain if I do the same for `getStorageElementName()` and `getChecksumType()` ?
Would you also require the backward compatibility for those that are probably not used anywhere but in DIRAC?
@chaen: what about `negociateProtocolWithOtherSE`?

I am fed up with all these unnecessary tests... and when we don't test `['OK']` we get rejected PRs ;-)

**Comment by phicharp** [comment]
I didn't know about that rule... IIRC Andrei once told me: if there is no other return but S_OK(), just return the result directly...
Otherwise we shall be forced to add 3 lines to test `result['OK']`, log a useless message and eventually return `result` because `myMethod()['Value']` will not be acceptable.
FYI there are already a lot of public methods in DIRAC that do NOT return `S_OK()/S_ERROR()`
Maybe a convention on the name of the method would be a possibility? I think we have far too many methods that start with `get`. IMHO `se.status()` would be sufficient and there is no need for `se.getStatus()`. So if we would say that `getStatus()` returns the shitty structure but `status()` returns the status, would it be  a possibility? 
I don't have a silver bullet, but what I know is that all this is a real pain! ... and don't tell me we should throw exceptions whenever there is an error ;-)!

**Comment by atsareg** [comment]
I do not think these changes are good. First, they change the interface which should never be done without a real good reason. Second, the SE status is evaluated using information from remote services and these queries can fail. In this particular case, the SE can fail instantiation most likely because of a failure to access the CS, or alike, and can be in an invalid state. In this case, I think the S_ERROR() should be returned instead of no any access which is rather mimicking a banned SE. 

As a general remark, any method which needs to access remote services must return S_OK/S_ERROR as the result depends on the things not under local control. Only methods that return some data from the object internal state can return values directly. This particular method returns response using only local data members, this is true. But the result can be still affected by a previous failed access to remote services as I said before, so we are back to the first case.

**Comment by phicharp** [comment]
@atsareg 
I think you missed the point here... The getStatus() method doesn't access ANY service and only returns S_OK(). This is what is being discussed... Whatever has happened before is totally irrelevant and there is no way it can return anything but a dictionary.
Instantiating a StorageElement cannot fail either as there is no way to fail the instantiation of an object. 
This is why the code is full of 
`StorageElement(seName).getStatus()['Value'][<item>]`
without any check on `['OK']`
Removing the `['Value']` is not a big deal unless you would like all these to be changed with explicit tests of `['OK']` that we know by construction will never fail... It is you who usually says one should not return the `S_OK/S_ERROR` structure if not needed... Do I miss anything?

**Comment by phicharp** [comment]
Added note:
`getStatus()` cannot call `isValid()` as `isValid()` calls `getStatus()`. Therefore if the user wants to test if an SE is valid, it should instantiate it first, check if it is valid and then call `getStatus()`. This is totally impractical, and one must only know that there is a potential "danger" to call directly `getStatus()` which is that the values are meaningless if something went wrong during instantiation. It is totally impossible to do that everywhere without severely affecting the code.

**Comment by phicharp** [comment]
I will propose a new implementation with 2 methods:

1. `se.status()` that returns a dictionary
2. `se.getStatus()` that calls first `isValid()` and returns `S_OK/S_ERROR`

Depending on where it is called, i.e. whether the validity is important we can then use either of them...

**Comment by atsareg** [comment]
The reason for these changes is that in the code there are many cases where the returned S_OK is not checked and this appears annoying ? I think that  this check simply must always be done if even it seems to you redundant. It is not "by construction" that it always returns S_OK, it is just in the current implementation done so. In some future implementation this can change. As I said before, this object represents a client accessing remote services and its state in general is defined by the state of those services. The fact that this state is evaluated once at construction and then served from a local data member is purely an implementation detail ( not correct in my view ). If your only goal is simplification of the code using this object, then it is not worth doing. Standard check of S_OK values adds few more lines indeed but these are absolutely understandable lines for a reader of the code. Standard patterns are not complicating the code. Introduction of a new method which does the same thing but returns a different structure is much more confusing for a developer who is not aware of this discussion.

**Comment by phicharp** [comment]
Andrei, this is how DIRAC had been using that method so far, not my invention! This method is assuming everywhere that it returns `S_OK()`
With the commit I just uploaded this is no longer true as it may return `S_ERROR` if the SE is invalid. I think this is what you wanted.
At some places however (like in the SE itself) we know that the SE is already valid because it is checked before, or we just would be happy with the default value `False` and then there is no need to check for the validity.

File: WorkloadManagementSystem/Executor/InputData.py [hunk 1]
```diff
@@ -234,53 +234,6 @@ def __checkReplicas( self, jobState, replicaDict, vo ):
     return self.__getSiteCandidates( okReplicas, vo )
 
   #############################################################################
-  def __checkActiveSEs( self, jobState, replicaDict ):
```
Comment (phicharp) [inline WorkloadManagementSystem/Executor/InputData.py]: This protected method is not used

**Comment by atsareg** [comment]
This is fine that getStatus() can return S_ERROR. As for the status(), how do you know that in a given context only internal local object status info is relevant ? Imagine that an object is created with certain access rules acquired from the CS/RSS/etc , but the rules are changed while the object still exists. Then it means that the object does not correspond any more to the current situation. This is true that with the current code it is never the case, but this can change in the future to take into account the situation described above.

**Comment by phicharp** [comment]
Can we see in due time? 
If one wants to check whether an SE is a `TapeSE` and the test is done in a positive way and `status()` returns `False` in case of problem, isn't it sufficient? This is anyway what all the users of `getStatus()` had been assuming. Therefore the proposed implementation does exactly the same as before, even adding a check that I think was missing ;-)

**Comment by chaen** [comment]
Review OK

**Comment by fstagni** [comment]
To me this would get a "review OK", how @atsareg  and @andresailer  review it?

**Pull request merged**